### PR TITLE
Porting tests to pytest

### PR DESCRIPTION
**Pull request by abadger**
* Achievement Get: No longer need mock/generator.py
  * Now done via pytest's parametrization
  * Port safe_eval to pytest
  * Port text tests to pytest
  * Port test_set_mode_if_different to pytest

* Move the AnsibleModules fixtures to module_utils/conftest.py for sharing

* More porting to obsolete mock/procenv.py
  * Port run_command to pytest
  * Port known_hosts tests to pytest
  * Port safe_eval to pytest
  * Remove unneeded import of procenv in test_postgresql

* Port test_pip to pytest style
  * As part of this, create a pytest ansiblemodule fixture in
    modules/conftest.py.  This is slightly different than the
    approach taken in module_utils because here we need to override the
    AnsibleModule that the modules will inherit from instead of one that
    we're instantiating ourselves.

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME
Various unittests

##### ANSIBLE VERSION

```
devel
```

### CODE REVIEW

**Comment by abadger** [comment]
\CC @mattclay @gundalow @pilou- 

(@pilou- CC'ing you because this has some more examples of porting from unittest to pytest)

File: test/units/modules/packaging/language/test_pip.py [hunk 1]
```diff
@@ -1,23 +1,18 @@
...
+def test_failure_when_pip_absent(mocker):
+    get_bin_path = mocker.patch('ansible.module_utils.basic.AnsibleModule.get_bin_path')
+    get_bin_path.return_value = None
 
-        mock_get_bin_path.return_value = None
-
-        with self.assertRaises(AnsibleFailJson):
-            set_module_args({'name': 'six'})
-            pip.main()
+    with pytest.raises(SystemExit):
```
Comment (pilou-) [inline test/units/modules/packaging/language/test_pip.py]: Both [`exit_json`](https://github.com/ansible/ansible/blob/ebd08d2a01d96d41f8bc1ccf27bc1bbf91060a44/lib/ansible/module_utils/basic.py#L2254) and [`fail_json`](https://github.com/ansible/ansible/blob/ebd08d2a01d96d41f8bc1ccf27bc1bbf91060a44/lib/ansible/module_utils/basic.py#L2261) methods call `sys.exit`: the old code checked that `fail_json` has been called and would have failed if `exit_json` were called in `pip.main`.

**Comment by abadger** [comment]
Okay, if that's what it's doing them we should be checking it a different
way.

Use the capfd fixture and then capfd.readouterr() and then assert that
failed true is in the Json that was printed to stdout.

We might also want to add a lot to this test.  Currently it is an
integration test using our unittesting framework.  In the future we might
want to have actual unittests here.

**Comment by abadger** [comment]
@pilou- Okay.  test_pip.py updated.

**Review by mattclay — APPROVED** [review]

**Comment by abadger** [comment]
Thanks @mattclay thanks @pilou-  !

**Pull request merged**